torch/sparse/_triton_ops.py

```diff
@@ -2354,7 +2355,7 @@ def _bsr_strided_addmm_kernel(
             # do block mm
             output_acc_block += tl.dot(
                 values_block, dense_block, allow_tf32=allow_tf32, out_dtype=acc_dtype
-            )
+            ).to(acc_dtype)
```

Comment:
Is this wrong?

Reply:
Hm, it should be redundant since `out_dtype` is set to `acc_dtype`.

Reply:
For `int8` the `acc_dtype` should be `int32`.

Reply:
It should, however, without this conversion, triton compile fails here with

<details>

```
Traceback (most recent call last):
  File "/home/pearu/git/pytorch/pytorch/test/test_sparse_csr.py", line 4081, in test_triton_kernel
    result = operation(*args, **kwargs)
  File "/home/pearu/git/pytorch/pytorch/torch/sparse/_triton_ops.py", line 1202, in bsr_dense_addmm
    launch_kernel(kernel, tensor_dims_map, full_grid, grid_blocks)
  File "/home/pearu/git/pytorch/pytorch/torch/sparse/_triton_ops.py", line 182, in launch_kernel
    kernel(grid, *sliced_tensors)
  File "/home/pearu/git/pytorch/pytorch/torch/sparse/_triton_ops.py", line 1187, in kernel
    _bsr_strided_addmm_kernel[grid](
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/site-packages/triton/runtime/jit.py", line 180, in <lambda>
    return lambda *args, **kwargs: self.run(grid=grid, warmup=False, *args, **kwargs)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/site-packages/triton/runtime/jit.py", line 401, in run
    self.cache[device][key] = compile(
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/site-packages/triton/compiler/compiler.py", line 263, in compile
    module = src.make_ir(options, context)
  File "/home/pearu/miniconda3/envs/pytorch-cuda-dev/lib/python3.9/site-packages/triton/compiler/compiler.py", line 108, in make_ir
    return ast_to_ttir(self.fn, self, context=context, options=options)
triton.compiler.errors.CompilationError: at 137:4:
    )

    # alpha is never 0
    if beta_is_nonzero:
        output_acc_block = tl.load(input_ptrs).to(acc_dtype)  # type: ignore[possibly-undefined]
        if not (beta_is_one and alpha_is_one):
            beta_alpha = beta / alpha
            output_acc_block *= beta_alpha
    else:
        output_acc_block = tl.zeros((BLOCKSIZE_ROW, BLOCKSIZE_COL), dtype=acc_dtype)

    for _ in range(row_nnz):
    ^
AssertionError('Loop-carried variable output_acc_block has initial type <[32, 16], int8> but is re-assigned to <[32, 16], int32> in loop! Please make sure that the type stays consistent.')
```

</details>

indicating that `tl.dot(..., out_dtype=torch.int8)` actually returns `int32` tensor.

If https://github.com/pytorch/pytorch/pull/133855#discussion_r1722285305 is what we want, `.to(acc_dtype)` should be removed (and find another fix to the assertion error), or `acc_dtype` must be defined as `torch.int32` for int8 inputs. WDYT?

Reply:
I assume this error stems from `output_acc_block` being allocated as `int8` initially and then `int32` (or acc_dtype) later on implicitly via the use of `out_dtype` in tl.dot.

Reply:
> I assume this error stems from output_acc_block being allocated as int8 initially

correct

> and then int32 (or acc_dtype) later on implicitly via the use of out_dtype in tl.dot.

`out_dtype` to `tl.dot` was `torch.int8` when the error occured. It is unexpected and likely a triton bug (I am using triton 3.0.0+989adb9a29).

Reply:
`.to(acc_dtype)` is now removed by using `acc_dtype=int32` when inputs are int8 tensors.